reps/2023-10-30-gpu-memory-support.md

```diff
@@ -0,0 +1,274 @@
+@ray.remote(gpu_memory=1024 * 1024 * 1024) # 1 mb request
+def task:
+  …
+```
+
+When a Ray node is started, Ray will auto detect the number of GPUs and GPU memory of each GPU and set the GPU and gpu_memory resources accordingly. Users also have the option to manually specify `gpu_memory` resources as the sum of total gpu memory across all gpus in the node. The default value is `num_gpus` * total memory of the gpu type in the node.
+
+```bash
+ray start # auto detection
+
+ray start –num_gpus=3 –gpu_memory=3000 * 1024 * 1024 * 1024 # manual override, each gpu has 1000mb total memory
+```
+
+Note that GPU memory and compute unit is 1-1 conversion, means 20GB of gpu memory is equivalent to 0.5 fractional value of an `A100_40GB` gpu. So, for simplicity and consistency, ray doesn't allow users to specify both `num_gpus` and `gpu_memory` in a single ray task/actor.
+
+```python
+# Request a fractional GPU both num_gpus and gpu_memory is not allowed
+@ray.remote(gpu_memory=1024 * 1024 * 1024, num_gpus=0.5) # raise ValueError exception
+def not_allowed_task:
+  …
+
+# Request a fractional GPU with specified num_gpus.
+@ray.remote(num_gpus=0.5)
+def num_gpus_task:
+  …
+
+# Request a fractional GPU with specified gpu_memory.
+@ray.remote(gpu_memory=1024 * 1024 * 1024) 
+def gpu_memory_task:
+  …
+```
+
+Additionally, users can still specify which gpu type they want to use by specifying `accelerator_type`.
+
+```python
+# Request a fractional of A100 GPU with specified gpu_memory
+@ray.remote(gpu_memory=1024 * 1024 * 1024 * 1024, accelerator_type="NVIDIA_A100")
+def nvidia_a100_gpu_task:
+  …
+
+# Requesting 30GB of gpu memory from a A10 GPU with 24GB of memory.
+# Task won't be able to be scheduled.
+@ray.remote(gpu_memory=30 * 1024 * 1024 * 1024 * 1024, accelerator_type="NVIDIA_TESLA_A10G")
+def nvidia_a10_gpu_task:
+  …
+```
+
+#### Placement Group
+TBD
+
+### Implementation
+The primary implementation entails the automatic detection of GPU memory during the initialization of a Ray cluster.
+
+```python
+class AcceleratorManager:
+    # return 0 if accelerator is not GPU, 
+    # else return total GPU memory of a single GPU
+    def get_current_node_gpu_memory(self):
+      ...  
+``` 
+
+Subsequently, we added another change within the scheduler to convert `gpu_memory` to `num_gpus` depending on which node the request got assigned to check if resource request is feasible in the node and allocate the resource.
+
+```pseudocode
+for node in node_list: 
+
+++  def convert_relative_resources(resource_request, node): # option 1 implementation
+++      if gpu_memory in resource_request: 
+++          resource_request.num_gpus = roundup(gpu_memory / node.label["gpu_memory_per_gpu"] , 0.0001)
+++          resource_request.gpu_memory = 0
+++      return resource_request
+++  convert_relative_resources(resource_request, node)
+
+    if check_is_feasible(resource_request):
+        allocation = TryAllocate(resource_request)
+```
```

Comment:
```suggestion
```pseudocode
for node in nodes: 
  def convert_relative_resources(resource_request, node):
      if gpu_memory in resource_request: 
          resource_request.num_gpus = roundup(gpu_memory / node.label["gpu_memory_per_gpu"] , 0.0001)
          resource_request.gpu_memory = 0
      return resource_request
 
  convert_relative_resources(resource_request, node)
 
  # After converting from gpu_memory to num_gpus, the remaining is the same  
  if check_is_available(resource_request, node):
        allocation = allocate(resource_request, node)
        break
```
```